src/main/java/com/larscheidschmitzhermes/nexus3/github/oauth/plugin/api/GithubApiClient.java

```diff
@@ -58,6 +69,22 @@ public GithubApiClient(GithubOauthConfiguration configuration) {
     }
 
     public GithubPrincipal authz(String login, char[] token) throws GithubAuthenticationException {
+        // Combine the login and the token as the cache key since they are both used to generate the principal. If either changes we should obtain a new
+        // principal.
+        String cacheKey = login + "|" + new String(token);
+        GithubPrincipal cached = tokenToPrincipalCache.getIfPresent(cacheKey);
+        if (cached != null) {
+            LOGGER.debug("Using cached principal for login: {}", login);
+            return cached;
+        } else {
+            GithubPrincipal principal = doAuthz(login, token);
+            tokenToPrincipalCache.put(cacheKey, principal);
+            return principal;
+        }
+    }
+
+    private GithubPrincipal doAuthz(String login, char[] token) throws GithubAuthenticationException {
```

Comment:
yes, the method was ugly already. We can probably clean that up after we merged.